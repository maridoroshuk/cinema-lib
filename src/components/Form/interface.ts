import { FormikHelpers } from 'formik';
import { ReactNode } from 'react';

export interface IFormProps<T> {
  children: ReactNode;
  initialValues: T;
  validate: (values: T) => void;
  onFormSubmit: (submitValues: T, { setSubmitting }: FormikHelpers<T>) => void;
}
