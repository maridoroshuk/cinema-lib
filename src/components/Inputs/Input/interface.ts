export interface IInputProps {
  type: string;
  name: string;
  label?: string;
  icon?: string;
  placeholder?: string;
  helperText?: string;
  dataCy: string;
}
