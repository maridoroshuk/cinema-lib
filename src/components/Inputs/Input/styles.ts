import { Field } from 'formik';
import styled from 'styled-components';

export const Wrapper = styled.div`
  background-color: ${({ theme }) => theme.colors.neutral900};
  padding: ${({ theme }) => `${theme.spaces[30]}px 0`};

  color: ${({ theme }) => theme.colors.neutral000};
`;

export const Label = styled.label<{ icon?: string }>`
  width: 100%;
  position: relative;
  display: inline-block;
  font-family: ${({ theme }) => theme.fontFamily.Poppins};
  font-size: ${({ theme }) => theme.sizes.fontSize[20]}px;
  word-break: break-word;
`;

export const StyledField = styled(Field)<{
  error: boolean;
  icon?: string;
}>`
  width: 100%;
  background-color: ${({ theme }) => theme.colors.neutral900};
  border: none;
  outline: none;
  border-bottom: ${({ theme, error }) =>
    `2.5px solid ${error ? theme.colors.danger : theme.colors.neutral000}`};
  color: ${({ theme, error }) =>
    error ? theme.colors.danger : theme.colors.neutral000};
  padding: ${({ theme }) => `${theme.spaces[15]}px 0`};
  font-family: ${({ theme }) => theme.fontFamily.Poppins};
  font-size: ${({ theme }) => theme.sizes.fontSize[20]}px;

  ::placeholder {
    color: ${({ theme }) => theme.colors.neutral000};
  }

  @media (min-width: ${({ theme }) => theme.screenSizes[900]}px) {
    max-width: ${({ theme }) => theme.sizes.width[630]}px;
  }
`;

export const Icon = styled.img`
  display: none;
  width: ${({ theme }) => theme.sizes.width[50]}px;
  height: ${({ theme }) => theme.sizes.width[50]}px;
  object-fit: cover;
  margin-right: ${({ theme }) => theme.spaces[30]}px;
  vertical-align: bottom;
  aspect-ratio: 1 / 1;

  @media (min-width: ${({ theme }) => theme.screenSizes[860]}px) {
    display: inline-block;
  }
`;

export const AdditionalText = styled.p`
  position: absolute;
  left: 0;
  font-style: italic;
  font-family: ${({ theme }) => theme.fontFamily.Poppins};
  font-size: ${({ theme }) => theme.sizes.fontSize[15]}px;
  word-break: break-word;
`;
