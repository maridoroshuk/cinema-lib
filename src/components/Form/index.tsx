import React from 'react';
import { Formik } from 'formik';

import { IFormProps } from './interface';
import { StyledForm } from './styles';

export const Form = <T extends {}>({
  children,
  initialValues,
  validate,
  onFormSubmit,
}: IFormProps<T>) => {
  return (
    <Formik
      initialValues={initialValues}
      validate={validate}
      onSubmit={onFormSubmit}
    >
      <StyledForm>{children}</StyledForm>
    </Formik>
  );
};
