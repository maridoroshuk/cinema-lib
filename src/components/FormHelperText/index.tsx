import React from 'react';

import { IFormHelperTextProps } from './interface';
import { Wrapper } from './styles';

export const FormHelperText = ({ message }: IFormHelperTextProps) => {
  return <Wrapper>{message}</Wrapper>;
};
