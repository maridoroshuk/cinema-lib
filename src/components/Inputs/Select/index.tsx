import React from 'react';

import { ISelectProps } from './interface';
import { Label, Option, StyledSelect, Wrapper } from './styles';

export const Select = ({
  config: { name, label, id, options },
}: ISelectProps) => {
  return (
    <Wrapper>
      <Label htmlFor={id}>
        {label}
        <StyledSelect component="select" name={name} data-testid="select">
          {options.map(({ id, value }) => (
            <Option key={id} value={value}>
              {value}
            </Option>
          ))}
        </StyledSelect>
      </Label>
    </Wrapper>
  );
};
