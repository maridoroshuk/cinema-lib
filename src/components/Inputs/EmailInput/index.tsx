import React, { useId } from 'react';
import { ErrorMessage } from 'formik';

import Send from '@/assets/icons/send.png';
import { Tooltip } from '@/components/Tooltip';

import { Icon, Label, StyledField, SubmitButton, Wrapper } from './styles';
import { IEmailInput } from './interface';

export const EmailInput = ({placeholder}: IEmailInput) => {
  const emailId = useId();

  return (
    <Wrapper>
      <Label htmlFor="email">
        <StyledField
          id={emailId}
          type="email"
          name="email"
          placeholder={placeholder}
          autoComplete="off"
        />
        <ErrorMessage name="email">
          {(message) => <Tooltip>{message}</Tooltip>}
        </ErrorMessage>
      </Label>
      <SubmitButton type="submit">
        <Icon src={Send} alt="send email" />
      </SubmitButton>
    </Wrapper>
  );
};
