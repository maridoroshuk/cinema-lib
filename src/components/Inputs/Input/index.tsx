import React, { useId } from 'react';
import { ErrorMessage } from 'formik';

import { Tooltip } from '@/components/Tooltip';

import { IInputProps } from './interface';
import { AdditionalText, Icon, Label, StyledField, Wrapper } from './styles';

export const Input = ({
  type,
  name,
  icon,
  label,
  placeholder,
  helperText,
  onInputBlur,
}: IInputProps) => {
  const inputId = useId();

  return (
    <Wrapper>
      <Label htmlFor={name}>
        {label}
        {icon && <Icon src={icon} alt={name} />}
        <StyledField
          id={inputId}
          type={type}
          name={name}
          placeholder={placeholder}
          autoComplete="off"
          onBlur={onInputBlur}
        />
        <ErrorMessage name={name}>
          {(message) => <Tooltip>{message}</Tooltip>}
        </ErrorMessage>
        {helperText && <AdditionalText>{helperText}</AdditionalText>}
      </Label>
    </Wrapper>
  );
};
