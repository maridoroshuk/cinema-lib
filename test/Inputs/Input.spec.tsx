import React from 'react';
import { render } from '@testing-library/react';

import Account from '@/assets/icons/account.png';
import { Form } from '@/components/Form';
import { Input } from '@/components/Inputs/Input';

import 'jest-styled-components';

import { setupTestProvider } from '../setupTests';

describe('Input', () => {
  const handleValidate = jest.fn();
  const handleSubmit = jest.fn();

  it('renders text input with icon', () => {
    const { container } = render(
      setupTestProvider(
        <Form
          initialValues={{ name: '' }}
          validate={handleValidate}
          onFormSubmit={handleSubmit}
        >
          <Input
            type="text"
            name="name"
            placeholder="Enter your name"
            icon={Account}
          />
        </Form>,
      ),
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveFormValues({ name: '' });
  });
  it('renders password input without icon', () => {
    const { container } = render(
      setupTestProvider(
        <Form
          initialValues={{ password: '' }}
          validate={handleValidate}
          onFormSubmit={handleSubmit}
        >
          <Input
            type="password"
            name="password"
            placeholder="Enter your password"
          />
        </Form>,
      ),
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveFormValues({ password: '' });
  });
});
