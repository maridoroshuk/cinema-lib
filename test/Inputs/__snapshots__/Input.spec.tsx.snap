// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders password input without icon 1`] = `
.c0 {
  width: 100%;
}

.c1 {
  background-color: #1E1F27;
  padding: 30px 0;
  color: #FFFFFF;
}

.c2 {
  width: 100%;
  position: relative;
  display: inline-block;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
  word-break: break-word;
}

.c3 {
  width: 100%;
  background-color: #1E1F27;
  border: none;
  outline: none;
  border-bottom: 2.5px solid #FFFFFF;
  color: #FFFFFF;
  padding: 15px 0;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
}

.c3::-webkit-input-placeholder {
  color: #FFFFFF;
}

.c3::-moz-placeholder {
  color: #FFFFFF;
}

.c3:-ms-input-placeholder {
  color: #FFFFFF;
}

.c3::placeholder {
  color: #FFFFFF;
}

@media (min-width:860px) {
  .c3 {
    max-width: 630px;
  }
}

<form
  action="#"
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="password"
    >
      <input
        autocomplete="off"
        class="c3"
        id="password"
        name="password"
        placeholder="Enter your password"
        type="password"
        value=""
      />
    </label>
  </div>
</form>
`;

exports[`Input renders text input with icon 1`] = `
.c0 {
  width: 100%;
}

.c1 {
  background-color: #1E1F27;
  padding: 30px 0;
  color: #FFFFFF;
}

.c2 {
  width: 100%;
  position: relative;
  display: inline-block;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
  word-break: break-word;
}

.c3 {
  width: 100%;
  background-color: #1E1F27;
  border: none;
  outline: none;
  border-bottom: 2.5px solid #FFFFFF;
  color: #FFFFFF;
  padding: 15px 0;
  font-family: 'Poppins',sans-serif;
  font-size: 20px;
}

.c3::-webkit-input-placeholder {
  color: #FFFFFF;
}

.c3::-moz-placeholder {
  color: #FFFFFF;
}

.c3:-ms-input-placeholder {
  color: #FFFFFF;
}

.c3::placeholder {
  color: #FFFFFF;
}

@media (min-width:860px) {
  .c3 {
    max-width: 630px;
  }
}

<form
  action="#"
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="name"
    >
      <input
        autocomplete="off"
        class="c3"
        id="name"
        name="name"
        placeholder="Enter your name"
        type="text"
        value=""
      />
    </label>
  </div>
</form>
`;
